### PR TITLE
Enable GDAX / Coinbase 

### PR DESCRIPTION
**Pull request by creslinux**
FT was sending 18 decimal places on buy/sell amounts.
Not all exchages support this precision.

CCXT provides the precision each pair supports.

This change makes use of that information prior to buy or sell.

Thank you for sending your pull request. But first, have you included
unit tests, and is your code PEP8 conformant? [More details](https://github.com/freqtrade/freqtrade/blob/develop/CONTRIBUTING.md)

Trades made by FT 

screen shot 2018-07-27 at 5 09 22 pm

### CODE REVIEW

File: freqtrade/exchange/__init__.py [hunk 1]
```diff
@@ -150,6 +154,18 @@ def exchange_has(self, endpoint: str) -> bool:
...
 
+    def symbol_prec(self, pair, amount: float):
+        '''
+        Returns the amount to buy or sell to the precision the Exchange accepts
+        :param amount: amount
+        :return: amount
+        '''
+        if self._api.markets[pair]['precision']['amount']:
+            symbol_prec = self._api.markets[pair]['precision']['amount']
+            amount = round(amount, symbol_prec)
```
[the remark this answers is not in the document]
Comment (creslinux) [reply freqtrade/exchange/__init__.py]: wouldn't that imply there was an amount on the exchange at a precision the exchange does not support we're round up?

**Comment by xmatthias** [comment]
if GDAX works flawlessly, we should add it to the list of supported exchanges i guess

but i assume will probably need more testing after this PR.

File: freqtrade/exchange/__init__.py [hunk 2]
```diff
@@ -167,6 +183,9 @@ def buy(self, pair: str, rate: float, amount: float) -> Dict:
             return {'id': order_id}
 
         try:
+            # Set the precision accepted by the exchange
+            amount = self.symbol_prec(pair, amount)
+
             return self._api.create_limit_buy_order(pair, amount, rate)
```
[the remark this answers is not in the document]
Comment (creslinux) [reply freqtrade/exchange/__init__.py]: Good spot, 
Precision is offered on price("rate") as well as amount. 

I'll wrap this in, look for a round-down too as revisiting. 

screen shot 2018-07-29 at 11 52 47 am

File: freqtrade/exchange/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (creslinux) [reply freqtrade/exchange/__init__.py]: I'll wrap this in as going to look at rounding(or not round) price too. 
I guess price should round up , amount sell down.

[the remark this answers is not in the document]
Comment (xmatthias) [reply freqtrade/exchange/__init__.py]: not necessarily, this depends on how the exchange handles fees (if it subtracts a fraction of a fee i can immagine this to happen).

**Comment by creslinux** [comment]
Precision is per symbol, then for amount and price. Which did you see 3 on?

Re relying on - we're currently broken on GDAX/coinbase, arguably the
largest by user exchange, easily 1st or 2nd global exchange.
Id take the stance we're moving from a broken position to a fixed one,
leveraging CCXT for its purpose to help here.

**Comment by xmatthias** [comment]
binance - ETH/BTC amount i belive ... 

for sure ccxt is not perfect -  i don't argue to work around issues there if we must do so.
all i meant was to not 100% trust the dict to be there for all exchanges - i've seen some values to be None too (was a while ago - and no, i don't remember which exchange) :(

File: freqtrade/exchange/__init__.py [hunk 3]
```diff
@@ -154,18 +151,38 @@ def exchange_has(self, endpoint: str) -> bool:
...
         '''
-        Returns the amount to buy or sell to the precision the Exchange accepts
+        Returns the amount to buy or sell to a precision the Exchange accepts
+        Rounded down
+        :param pair: the symbol
         :param amount: amount
         :return: amount
         '''
-        if self._api.markets[pair]['precision']['amount']:
+        if self._api.markets[pair]['precision']['amount'] > 0:
```
[the remark this answers is not in the document]
Comment (creslinux) [reply freqtrade/exchange/__init__.py]: I don't agree with this argument, we could say the same for any field we rely on from CCXT.

Currently Freqtrade supports just 1 of the top 15 exchanges, then only as binance is handling our bad behaviour.

The majority of binance buy amounts request an amount precision between 2 and 8, we send 18....   We could argue if binance tomorrow enforced their precision we would be broken as not operating to spec. 

screen shot 2018-07-29 at 8 05 44 pm

**Comment by creslinux** [comment]
0 is valid on many 
If im buying a shitcoin that is 1,234,456 for 1 BTC, binance wont allow me to buy a fraction of an amount.

**Comment by creslinux** [comment]
Here is where 0 is a valid precision. 
The pretty dict is direct exchange info from binance and the debug CCXT

ICN has a min order of 1.0 and a step size of 1.  This means any fractions of orders are invalid. 

screen shot 2018-07-29 at 8 20 30 pm

**Comment by creslinux** [comment]
An out of date human table on amount and price precision from binance. 
but depicts there trade rules for precision well 

https://support.binance.com/hc/en-us/articles/115000594711-Trading-Rule

**Comment by xmatthias** [comment]
as long as we're aware ... that's fine 👍 
i agree, this is prone to happen at the moment anyway (but was also not the point i was trying to make) - i was just ... questioning the correctness, but your link clarified that

**Comment by creslinux** [comment]
Awesome - lets get GDAX online?

File: freqtrade/exchange/__init__.py [hunk 4]
```diff
@@ -150,6 +151,38 @@ def exchange_has(self, endpoint: str) -> bool:
...
+        '''
+        Returns the amount to buy or sell to a precision the Exchange accepts
+        Rounded down
+        :param pair: the symbol
+        :param amount: amount
+        :return: amount
+        '''
+        if self._api.markets[pair]['precision']['amount']:
+            symbol_prec = self._api.markets[pair]['precision']['amount']
+            multiplier = int('1' + ('0' * symbol_prec))
```
Comment (xmatthias) [inline freqtrade/exchange/__init__.py]: could you use `pow(10, symbol_prec)`? 
that would also give you 1000 without playing string-tricks ...

**Comment by xmatthias** [comment]
you won't like me for the next thing

tests ... ?

**Comment by creslinux** [comment]
The tests showing were sending bad data to all exchanges are definitely
missing.

I'm going to focus my other work, drip feeding review is disproportionately
time consuming to me creating environments etc. It interupts what I'm
trying to deliver.

At the moment FT is broken on any exchange working to spec and only works
by luck on those that are forgiving of our bad data input.

Fixing an incident should take priority over not in my view of the world.

**Comment by xmatthias** [comment]
you say it's broken - but GDAX is not a supported exchange until now, and for binance the CCXT rounding seems to work just fine.

so i don't see this as a "superurgent" incident unless proven otherwise, sorry.

**Comment by creslinux** [comment]
Screw up on GIT - creating a new branch for the PR

**Pull request closed without merging**